src/helper-x11/xorguserhelper.cpp

```diff
@@ -0,0 +1,274 @@
+
+namespace SDDM {
+
+StartupEvent::StartupEvent()
+    : QEvent(StartupEventType)
+{
+}
+
+XOrgUserHelper::XOrgUserHelper(int fd, const QString &serverCmd, const QString &clientCmd,
+                               QObject *parent)
+    : QObject(parent)
+    , m_fd(fd)
+    , m_serverCmd(serverCmd)
+    , m_clientCmd(clientCmd)
+{
+    m_xauth.setAuthDirectory(QStandardPaths::writableLocation(QStandardPaths::RuntimeLocation));
+}
+
+bool XOrgUserHelper::start()
+{
+    // Create xauthority
+    m_xauth.setup();
+
+    // Start server process
+    if (!startServer())
+        return false;
+
+    // Setup display
+    startDisplayCommand();
+
+    // Start client process
+    if (!startClient())
+        return false;
+
+    return true;
+}
+
+void XOrgUserHelper::stop()
+{
+    if (m_clientProcess) {
+        qInfo("Stopping client...");
+        m_clientProcess->terminate();
+        if (!m_clientProcess->waitForFinished(5000))
+            m_clientProcess->kill();
+        m_clientProcess->deleteLater();
+        m_clientProcess = nullptr;
+    }
+
+    if (m_serverProcess) {
+        qInfo("Stopping server...");
+        m_serverProcess->terminate();
+        if (!m_serverProcess->waitForFinished(5000))
+            m_serverProcess->kill();
+        m_serverProcess->deleteLater();
+        m_serverProcess = nullptr;
+
+        displayFinished();
+    }
+}
+
+void XOrgUserHelper::customEvent(QEvent *event)
+{
+    if (event->type() == StartupEventType) {
+        if (!start())
+            QCoreApplication::exit(EXIT_FAILURE);
+    }
+}
+
+bool XOrgUserHelper::startProcess(const QString &cmd,
+                                  const QProcessEnvironment &env,
+                                  QProcess **p)
+{
+    auto args = QProcess::splitCommand(cmd);
+    const auto program = args.takeFirst();
+
+    // Make sure to forward the input of sddm-helper-x11 onto the Xorg
+    // server, otherwise it will complain that only console users are allowed
+    auto *process = new QProcess(this);
+    process->setInputChannelMode(QProcess::ForwardedInputChannel);
+    process->setProcessChannelMode(QProcess::ForwardedChannels);
+    process->setProcessEnvironment(env);
+
+    connect(process, QOverload<int, QProcess::ExitStatus>::of(&QProcess::finished),
+            process, [](int exitCode, QProcess::ExitStatus exitStatus) {
+        if (exitCode != 0 || exitStatus != QProcess::NormalExit)
+            QCoreApplication::instance()->quit();
+    });
+
+    process->start(program, args);
+    if (!process->waitForStarted(10000)) {
+        qWarning("Failed to start \"%s\": %s",
+                 qPrintable(cmd),
+                 qPrintable(process->errorString()));
+        return false;
+    }
+
+    if (p)
+        *p = process;
+
+    return true;
+}
+
+bool XOrgUserHelper::startServer()
+{
+    // Create pipe for communicating with X server
+    // 0 == read from X, 1 == write to X
+    int pipeFds[2];
+    if (::pipe(pipeFds) != 0) {
```

Comment:
The file descriptor for the read endpoint will be leaked to the X server process. Maybe set the O_CLOEXEC flag after creating the pipe file descriptors?

```
fcntl(pipeFds[0], F_SETFD, FD_CLOEXEC);
```